proposals/types-as-annotation-arguments.md

```diff
@@ -0,0 +1,126 @@
+# Types as annotation arguments
+
+* **Type**: Design proposal
+* **Author**: Andrey Breslav
+* **Status**: Under consideration
+* **Prototype**: Not started
+
+Discussion of this proposal is held in [this issue](TODO).
+
+## Synopsis
+
+Currently, we can only mention classes in annotation arguments: `@Ann(Foo::class)`. Although in some contexts terms *class* and *type* can be used interchangeably, those are two different things, for example, a class `Foo` can be mentioned in many different types: `Foo`, `Foo<Bar>`, `Foo<Foo<Bar>>`, `Foo<*>`, `Bar<Foo>, etc.
+   
+Sometimes we need to refer to types, not classes. For example, in many cases it's useless to say `List::class`, we need more information like `List<String>` or `List<Nothing>`.
+    
+We propose adding support for types as annotation arguments (and, optionally, elsewhere in the language):
+- an annotation parameter may be declared to be of type `KType`,
+- an annotation argument may be a type (in some syntactic form, e.g. `@Ann(Foo<Bar>)`).
+
+
+Example (syntax is subject to discussion):
+
+``` kotlin
+annotation class Ann(val type: KType)
+
+@Ann(type = Foo<Bar>)
+fun test() {} 
+```
+
+## References
+
+- [KT-42](https://youtrack.jetbrains.com/issue/KT-42) Support type parameters for annotations
+- [KT-6563](https://youtrack.jetbrains.com/issue/KT-6563) Support reified type parameters on annotation classes
+
+## Syntax
+
+> A thing to keep in mind: the same syntax is likely to be relevant for otehr uses of types as values, e.g. for reflection literals
+
+### Strawman types: `@Ann(Foo<Bar>)` 
+
+- **pro**: the most intuitive syntax possible
+- **con**: as types are syntactically ambiguous with expressions (simples example: `Foo`), using types as they are may be problematic for parsing,
+
+### Reflection-like literals
+
+Something like `Foo<Bar>::type`.
+
+- **pro**: looks like reflection,
+- **con**: the type on the left may have a member named `type` (which is too popular to be made a keyword), so this syntax is ambiguous with a member reference.
+
+### `typeof`
+
+Something like `@Ann(typeof(Foo<Bar>))`. Note: `typeof` is [reserved as a keyword](https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/lexer/Kotlin.flex#L249) already.
```

Comment:
This link is dead. Permalinks should be used instead to prevent future occurrence.

Reply:
Thanks, fixed